dap_aria_mapping/utils/novelty_utils.py

```diff
@@ -0,0 +1,255 @@
+
+    Returns:
+        pd.DataFrame: A dataframe with columns "topic_1", "topic_2", "year", "counts"
+    """
+    return document_topic_pairs.groupby(
+        ["topic_1", "topic_2", "topic_pair", "year"], as_index=False
+    ).agg(counts=("work_id", "count"))
+
+
+def get_counts_of_pairs_with_topic(
+    document_topic_pairs: pd.DataFrame, id_column="work_id"
+) -> dict:
+    """
+    Calcuate all N_i_t / N_j_t values:
+    Given a table with document ids and topic pairs mentioned in the document,
+    calculates the number of pairs in which the topic pair appears
+
+    NB: The input dataframe should contain topic pairs only for a single year
+
+    Args:
+        document_topic_pairs (pd.DataFrame): Dataframe with columns "work_id", "topic_1", "topic_2", "year"
+    """
+    topic_counts_df = (
+        pd.concat(
+            [
+                document_topic_pairs[["topic_1", "counts"]],
+                document_topic_pairs[["topic_2", "counts"]],
+            ]
+        )
+        # Combine topic_1 and topic_2 into one column
+        .assign(topics=lambda df: df["topic_1"].fillna(df["topic_2"]))
+        # Count number of times each topic appears
+        .groupby("topics", as_index=False).agg(counts=("counts", "sum"))
+    )
+    # Convert to dict
+    return dict(zip(topic_counts_df["topics"], topic_counts_df["counts"]))
+
+
+def get_topic_pair_commonness(
+    document_topic_pairs: pd.DataFrame,
+    N_t: int,
+    topic_counts_dict: dict,
+    topic_pair_counts_dict: dict,
+):
+    """
+    Calculate the novelty score for each document
+
+    Args:
+        document_topic_pairs (pd.DataFrame): Dataframe with columns "work_id", "topic_1", "topic_2", "year"
+        N_t (int): Total number of documents
+        topic_counts_dict (dict): Dictionary with topic counts
+        topic_pair_counts_dict (dict): Dictionary with topic pair counts
+    """
+    # Calculate the novelty score for each document
+    return (
+        document_topic_pairs.assign(
+            N_ij_t=lambda df: df["topic_pair"].apply(
+                lambda x: topic_pair_counts_dict[x]
+            )
+        )
+        .assign(N_i_t=lambda df: df["topic_1"].apply(lambda x: topic_counts_dict[x]))
+        .assign(N_j_t=lambda df: df["topic_2"].apply(lambda x: topic_counts_dict[x]))
+        .assign(commonness=lambda df: df["N_ij_t"] * N_t / (df["N_i_t"] * df["N_j_t"]))
+    )
+
+
+def aggregate_document_novelty(
+    topic_pair_commonness: pd.DataFrame,
+    id_column: str = "work_id",
+) -> pd.DataFrame:
+    """
+    Calculate the novelty score for each document
+    """
+    return (
+        topic_pair_commonness.groupby([id_column, "year"], as_index=False)
+        .agg(commonness=("commonness", (lambda x: aggregate_document_commonness(x))))
+        .assign(novelty=lambda df: novelty_score(df["commonness"]))
+    )
+
+
+def novelty_score(commonness_score: float) -> float:
```

Comment:
this function name could be a bit more specific as it's not really calculating the full novelty score and I was a bit confused with `def document_novelty()`